### PR TITLE
Implement new dashboard APIs

### LINKED ISSUE #4451: Adapt upcoming events widget for mobile clients
### Is your feature request related to a problem? Please describe.

With https://github.com/nextcloud/server/pull/33658 mobile clients can access app widget data.

### Describe the solution you'd like

Adapt the *Upcoming events* widget to https://github.com/nextcloud/server/pull/33658 once it's done.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by julien-nc**
closes #4451 

requires https://github.com/nextcloud/server/pull/33658

In progress

### CODE REVIEW

**Comment by eneiluj** [comment]
@ChristophWurst I just realized Calendar supports NC >= 22. The new APIs will only be there from 25. How do you wanna deal with that? Bump the min NC version to 25?

**Comment by tcitworld** [comment]
Needed for https://github.com/nextcloud/calendar/pull/4459 anyway

**Comment by eneiluj** [comment]
@ChristophWurst I'm stuck in getting the next occurrence of the events instead of DTSTART. I don't want to implement an overkill method when something simple might already exist :grin:. Could you point me to places to understand how it's done by the frontend widget? Do I have to compute it from the event repetition rules?

**Comment by eneiluj** [comment]
@tcitworld True that, thanks for the remark.

File: lib/Dashboard/CalendarWidget.php [hunk 1]
```diff
@@ -97,14 +108,82 @@ public function getUrl(): ?string {
...
+			/** @var DateTimeImmutable $startDate */
+			$startDate = $event['objects'][0]['DTSTART'][0];
+			return new WidgetItem(
+				$event['objects'][0]['SUMMARY'][0] ?? '',
+				$this->dateTimeFormatter->formatTimeSpan(DateTime::createFromImmutable($startDate)),
+				// TODO fix this route and get the correct objectId
+				$this->urlGenerator->getAbsoluteURL(
+					$this->urlGenerator->linkToRoute('calendar.view.index', ['objectId' => $event['uid']])
+				),
+				// TODO find or implement and endpoint providing colored dots images
```
Comment (tobiasKaminsky) [inline lib/Dashboard/CalendarWidget.php:166]: If "round-icon" is provided any square, colored svg should be suitable :)

File: lib/Dashboard/CalendarWidget.php [hunk 2]
```diff
@@ -25,41 +25,52 @@
...
 namespace OCA\Calendar\Dashboard;
 
+use DateInterval;
+use DateTime;
+use DateTimeImmutable;
 use OCA\Calendar\AppInfo\Application;
 use OCA\Calendar\Service\JSDataService;
-use OCP\Dashboard\IWidget;
-use OCP\IInitialStateService;
+use OCA\DAV\CalDAV\CalDavBackend;
```
Comment (ChristophWurst) [inline lib/Dashboard/CalendarWidget.php:33]: private API of another app. we should be able to go through `\OCP\Calendar\IManager::searchForPrincipal`

**Comment by miaulalala** [comment]
@tobiasKaminsky please tell me if you can work with this response format:

```xml
 <data>
  <calendar>
   <element>
    <subtitle>in 5 days</subtitle>
    <title>Test</title>
    <link>http://nextcloud.local/index.php/apps/calendar/?objectId=59d30b6c-5a31-4d28-b1d6-c8f928180e96</link>
    <iconUrl>data:image/svg+xml;base64,PHN2ZyBoZWlnaHQ9IjMyIiB3aWR0aD0iMzIiIHZpZXdCb3g9IjAgMCAzMiAzMiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cmVjdCB3aWR0aD0iMTAwJSIgaGVpZ2h0PSIxMDAlIiBmaWxsPSIjOTc1MEE0Ii8+PC9zdmc+</iconUrl>
    <sinceId>1666051200</sinceId>
   </element>
   <element>
    <subtitle>in 8 days</subtitle>
    <title>terretrrt</title>
    <link>http://nextcloud.local/index.php/apps/calendar/?objectId=f56c7bbd-8f0a-4d88-b63e-dbeb76c72ec2</link>
    <iconUrl>data:image/svg+xml;base64,PHN2ZyBoZWlnaHQ9IjMyIiB3aWR0aD0iMzIiIHZpZXdCb3g9IjAgMCAzMiAzMiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cmVjdCB3aWR0aD0iMTAwJSIgaGVpZ2h0PSIxMDAlIiBmaWxsPSIjOTc1MEE0Ii8+PC9zdmc+</iconUrl>
    <sinceId>1666310400</sinceId>
   </element>
   <element>
    <subtitle>in 9 days</subtitle>
    <title>tesd</title>
    <link>http://nextcloud.local/index.php/apps/calendar/?objectId=1821edd2-7cb4-47a5-84c4-8d8a75c0db9f</link>
    <iconUrl>data:image/svg+xml;base64,PHN2ZyBoZWlnaHQ9IjMyIiB3aWR0aD0iMzIiIHZpZXdCb3g9IjAgMCAzMiAzMiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cmVjdCB3aWR0aD0iMTAwJSIgaGVpZ2h0PSIxMDAlIiBmaWxsPSIjMDA4MmM5Ii8+PC9zdmc+</iconUrl>
    <sinceId>1666396800</sinceId>
   </element>
   <element>
    <subtitle>in 10 days</subtitle>
    <title>rers</title>
    <link>http://nextcloud.local/index.php/apps/calendar/?objectId=d335eaea-6490-4968-936a-eccd32b04a3b</link>
    <iconUrl>data:image/svg+xml;base64,PHN2ZyBoZWlnaHQ9IjMyIiB3aWR0aD0iMzIiIHZpZXdCb3g9IjAgMCAzMiAzMiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cmVjdCB3aWR0aD0iMTAwJSIgaGVpZ2h0PSIxMDAlIiBmaWxsPSIjMDA4MmM5Ii8+PC9zdmc+</iconUrl>
    <sinceId>1666598400</sinceId>
   </element>
   <element>
    <subtitle>in 10 days</subtitle>
    <title>sdfsdfd</title>
    <link>http://nextcloud.local/index.php/apps/calendar/?objectId=750acdde-8eaa-4cac-ba25-50d79011945d</link>
    <iconUrl>data:image/svg+xml;base64,PHN2ZyBoZWlnaHQ9IjMyIiB3aWR0aD0iMzIiIHZpZXdCb3g9IjAgMCAzMiAzMiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cmVjdCB3aWR0aD0iMTAwJSIgaGVpZ2h0PSIxMDAlIiBmaWxsPSIjQzk4ODc5Ii8+PC9zdmc+</iconUrl>
    <sinceId>1666483200</sinceId>
   </element>
  </calendar>
</data>
```

the icon url is now a base64 encoded svg with the calendar colour.

File: lib/Dashboard/CalendarWidget.php (hunk 2 above)
Comment (ChristophWurst) [reply lib/Dashboard/CalendarWidget.php:33]: @miaulalala why collapse if this hasn't been addressed?

Comment (miaulalala) [reply lib/Dashboard/CalendarWidget.php]: sorry, didn't push the changes but had it done locally already. it's done now

**Comment by miaulalala** [comment]
Thinking about this some more I haven't properly implemented the limit param. Reworking tomorrow.

**Comment by tobiasKaminsky** [comment]
iconUrl must be a real URL :S

**Comment by tobiasKaminsky** [comment]
A dummy svg with replaced background color would work:
```
<svg width="64mm" height="64mm" viewBox="0 0 64 64" xmlns="http://www.w3.org/2000/svg">
    <path style="color:#000;overflow:visible;fill:red;stroke:none;stroke-width:0;paint-order:markers fill stroke" d="M0 0h64v64H0z"/>
</svg>
```

File: lib/Controller/ViewController.php [hunk 3]
```diff
@@ -148,4 +149,18 @@ private function getView(string $view): string {
 				return $view;
 		}
 	}
+
+	/**
+	 * @NoCSRFRequired
```
Comment (ChristophWurst) [inline lib/Controller/ViewController.php:154]: no admin required

Comment (miaulalala) [reply lib/Controller/ViewController.php:154]: instead or in addition?

File: lib/Controller/ViewController.php [hunk 4]
```diff
@@ -148,4 +149,18 @@ private function getView(string $view): string {
...
+	 *
+	 * This function makes the colour dots work for mobile widgets
+	 *
+	 * Returns an SVG with size32x32
+	 */
+	public function getCalendarDotSvg(string $color) {
+		$svg = '<svg height="32" width="32" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg"><rect width="100%" height="100%" fill="' . urldecode($color) . '"/></svg>';
```
Comment (tcitworld) [inline lib/Controller/ViewController.php:161]: Maybe some validation is better to make sure this is not abused? `calendar-color` *should* always be an hexadecimal color.

File: lib/Controller/ViewController.php (hunk 3 above)
Comment (ChristophWurst) [reply lib/Controller/ViewController.php:154]: either. I don't know how client auth works. they might or might not send a CSRF token.

there is also a chance that the URL doesn't work at all for clients. AFAIK they have to be OCS

File: lib/Controller/ViewController.php (hunk 4 above)
Comment (miaulalala) [reply lib/Controller/ViewController.php:161]: good point

**Comment by miaulalala** [comment]
Testing is possible with:
`curl -k -H 'OCS-APIRequest: true' -u username:password http://your_base_url/ocs/v2.php/apps/dashboard/api/v1/widget-items -v`

**Comment by ChristophWurst** [comment]
> Testing is possible

Nice. Could you test this icon URLs this way as a non-admin user?

**Comment by miaulalala** [comment]
> > Testing is possible
> 
> Nice. Could you test this icon URLs this way as a non-admin user?

will do

**Comment by miaulalala** [comment]
For my user bob who is not an admin it works well.

```xml
  <calendar>
   <element>
    <subtitle>in 4 days</subtitle>
    <title>fdsddsf</title>
    <link>http://nextcloud.local/index.php/apps/calendar/?objectId=e68c5668-c777-4069-bfc4-186113621b0d</link>
    <iconUrl>http://nextcloud.local/index.php/apps/calendar/public/getCalendarDotSvg/%230082c9</iconUrl>
    <sinceId>1666051200</sinceId>
   </element>
   <element>
    <subtitle>in 7 days</subtitle>
    <title>dsfdfsdfs</title>
    <link>http://nextcloud.local/index.php/apps/calendar/?objectId=fb1883be-abad-4b3e-ba4a-df92180d5fab</link>
    <iconUrl>http://nextcloud.local/index.php/apps/calendar/public/getCalendarDotSvg/%230082c9</iconUrl>
    <sinceId>1666310400</sinceId>
   </element>
   <element>
    <subtitle>in 10 days</subtitle>
    <title>fdsfddsffdsdfssdsdfsdfsfddsf</title>
    <link>http://nextcloud.local/index.php/apps/calendar/?objectId=48ec3b78-68f4-4278-be9d-799715fb04f1</link>
    <iconUrl>http://nextcloud.local/index.php/apps/calendar/public/getCalendarDotSvg/%230082c9</iconUrl>
    <sinceId>1666569600</sinceId>
   </element>
   <element>
    <subtitle>in 11 days</subtitle>
    <title>dssdadsasadasd</title>
    <link>http://nextcloud.local/index.php/apps/calendar/?objectId=f2980c2e-5f8f-4153-9a41-9f90f55c6d8a</link>
    <iconUrl>http://nextcloud.local/index.php/apps/calendar/public/getCalendarDotSvg/%23499AA2</iconUrl>
    <sinceId>1666656000</sinceId>
   </element>
   <element>
    <subtitle>in 12 days</subtitle>
    <title>dsadsasdasadsad</title>
    <link>http://nextcloud.local/index.php/apps/calendar/?objectId=bbd0e106-1473-4f0f-97ef-6270695616dc</link>
    <iconUrl>http://nextcloud.local/index.php/apps/calendar/public/getCalendarDotSvg/%23499AA2</iconUrl>
    <sinceId>1666742400</sinceId>
   </element>
  </calendar>

```

and the icons work and are in the right colour.

File: lib/Controller/ViewController.php (hunk 4 above)
Comment (tcitworld) [reply lib/Controller/ViewController.php:169]: FYI, to make sure I tested on a production database and only 0.004% of entries didn't match that. They were lacking the leading `#`, which probably meant a bugged 3rd-party client.

In any case I think the behavior should be to reset `$color` to a valid one, not sure about sending an exception like this.

File: lib/Dashboard/CalendarWidget.php [hunk 5]
```diff
@@ -97,14 +113,86 @@ public function getUrl(): ?string {
...
+		foreach ($calendars as $calendar) {
+			$searchResult = $calendar->search('', [], $options, $limitPerCalendar);
+			foreach ($searchResult as $calendarEvent) {
+				/** @var DateTimeImmutable $startDate */
+				$startDate = $calendarEvent['objects'][0]['DTSTART'][0];
+				$widget = new WidgetItem(
+					$calendarEvent['objects'][0]['SUMMARY'][0] ?? 'New Event',
+					$this->dateTimeFormatter->formatTimeSpan(DateTime::createFromImmutable($startDate)),
+					$this->urlGenerator->getAbsoluteURL($this->urlGenerator->linkToRoute('calendar.view.index', ['objectId' => $calendarEvent['uid']])),
+					$this->urlGenerator->getAbsoluteURL($this->urlGenerator->linkToRoute('calendar.view.getCalendarDotSvg', ['color' => $calendar->getDisplayColor()])),
```
Comment (tcitworld) [inline lib/Dashboard/CalendarWidget.php:175]: `getDisplayColor` can return `null` and the route doesn't allow the parameter to be empty.

**Review by tcitworld — CHANGES_REQUESTED** [review]

File: lib/Controller/ViewController.php (hunk 4 above)
Comment (miaulalala) [reply lib/Controller/ViewController.php:169]: yes another good point, thanks for testing on a live instance. I changed the logic to return a svg in any case, using the default nextcloud blue as a fallback.

File: css/dashboard.css [hunk 6]
```diff
@@ -0,0 +1,8 @@
+.app-icon-calendar {
+	background-image: url('../img/calendar-dark.svg');
+}
+
+/* for NC <= 24 */
+body.theme--dark .app-icon-calendar {
```
Comment (ChristophWurst) [inline css/dashboard.css:6]: we don't support <=24 -> :fire:

File: lib/Dashboard/CalendarWidget.php [hunk 7]
```diff
@@ -97,14 +113,86 @@ public function getUrl(): ?string {
...
+		}
+		return $widgetItems;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getWidgetButtons(string $userId): array {
+		return [
+			new WidgetButton(
```
Comment (ChristophWurst) [inline lib/Dashboard/CalendarWidget.php:189]: shouldn't there be a *New event* button if the widget is empty?

File: tests/php/unit/Controller/ViewControllerTest.php [hunk 8]
```diff
@@ -61,6 +62,9 @@ class ViewControllerTest extends TestCase {
 	/** @var ViewController */
 	private $controller;
 
+	/** @var IAppData|IAppData&MockObject|MockObject  */
```
Comment (ChristophWurst) [inline tests/php/unit/Controller/ViewControllerTest.php:65]:
```suggestion
	/** @var IAppData|MockObject  */
```

File: tests/php/unit/Dashbaord/CalendarWidgetTest.php [hunk 9]
```diff
@@ -23,33 +23,69 @@
...
 use OCP\IL10N;
+use OCP\IURLGenerator;
+use PHPUnit\Framework\MockObject\MockObject;
+use Safe\DateTimeImmutable;
+use Test\TestCase;
 
 class CalendarWidgetTest extends TestCase {
 
-	/** @var IL10N|\PHPUnit\Framework\MockObject\MockObject */
+	/** @var IL10N|IL10N&MockObject|MockObject  */
```
Comment (ChristophWurst) [inline tests/php/unit/Dashbaord/CalendarWidgetTest.php:42]: what is this `&` notation?

Comment (miaulalala) [reply tests/php/unit/Dashbaord/CalendarWidgetTest.php:42]: done automatically by my IDE, maybe a setting.

File: lib/Dashboard/CalendarWidget.php (hunk 7 above)
Comment (miaulalala) [reply lib/Dashboard/CalendarWidget.php:189]: I don't know.

Comment (miaulalala) [reply lib/Dashboard/CalendarWidget.php:189]: This needs some more thought as the route is named the same and then decided via postfix, which doesn't work with the current url generation. leave as a to do for later

**Comment by miaulalala** [comment]
@tobiasKaminsky pls review!

**Review by ChristophWurst — APPROVED** [review]

File: lib/Controller/ViewController.php [hunk 10]
```diff
@@ -148,4 +156,37 @@ private function getView(string $view): string {
...
+	 *
+	 * @param string $color - url encoded HEX colour
+	 * @return FileDisplayResponse
+	 * @throws NotFoundException
+	 * @throws NotPermittedException
+	 */
+	public function getCalendarDotSvg(string $color = "#0082c9"): FileDisplayResponse {
+		if (preg_match('/^#([0-9a-f]{3}|[0-9a-f]{6})$/i', urldecode($color))) {
+			$svg = '<svg height="32" width="32" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg"><rect width="100%" height="100%" fill="' . urldecode($color) . '"/></svg>';
+		} elseif (preg_match('/^#([0-9a-f]{3}|[0-9a-f]{6})$/i', '#' . urldecode($color))) {
```
Comment (tcitworld) [inline lib/Controller/ViewController.php:176]: I would only define `$validColor` in each condition branch, then define `$svg` only once below for clarity.

You could also have made the `#` in the regex optional, then `trim($color, '#')` and finally add the `#` in the svg template itself. That would save a `elseif` call, but probably not worth it in terms of performance.

**Review by tcitworld — APPROVED** [review]

**Pull request merged**